packages/firestore/src/api/database.ts

```diff
@@ -893,102 +877,129 @@ export class DocumentReference<T = PublicDocumentData>
-        }
-      },
-      error: args[currArg + 1] as ErrorFn,
-      complete: args[currArg + 2] as CompleteFn
-    };
-
-    return firestoreClientListen(
-      this._firestoreClient,
-      newQueryForPath(this._key.path),
-      internalOptions,
-      observer
+    const options = extractSnapshotOptions(args);
+    const observer = wrapObserver<DocumentSnapshot<T>, exp.DocumentSnapshot<T>>(
+      args,
+      result =>
+        new DocumentSnapshot(
+          this.firestore,
+          result._key,
+          result._document,
+          result.metadata.fromCache,
+          result.metadata.hasPendingWrites,
+          this._delegate._converter as UntypedFirestoreDataConverter<T>
+        )
     );
+    return onSnapshot(this._delegate, options, observer);
   }
 
   get(options?: PublicGetOptions): Promise<PublicDocumentSnapshot<T>> {
-    if (options && options.source === 'cache') {
-      return firestoreClientGetDocumentFromLocalCache(
-        this._firestoreClient,
-        this._key
-      ).then(
-        doc =>
-          new DocumentSnapshot(
-            this.firestore,
-            this._key,
-            doc,
-            /*fromCache=*/ true,
-            doc instanceof Document ? doc.hasLocalMutations : false,
-            this._converter
-          )
-      );
+    let snap: Promise<exp.DocumentSnapshot<T>>;
+    if (options?.source === 'cache') {
+      snap = getDocFromCache(this._delegate);
+    } else if (options?.source === 'server') {
+      snap = getDocFromServer(this._delegate);
     } else {
-      return firestoreClientGetDocumentViaSnapshotListener(
-        this._firestoreClient,
-        this._key,
-        options
-      ).then(snapshot => this._convertToDocSnapshot(snapshot));
+      snap = getDoc(this._delegate);
     }
+
+    return snap.then(
+      result =>
+        new DocumentSnapshot(
+          this.firestore,
+          result._key,
+          result._document,
+          result.metadata.fromCache,
+          result.metadata.hasPendingWrites,
+          this._delegate._converter as UntypedFirestoreDataConverter<T>
+        )
+    );
   }
 
   withConverter<U>(
     converter: PublicFirestoreDataConverter<U>
   ): PublicDocumentReference<U> {
-    return new DocumentReference<U>(this._key, this.firestore, converter);
+    return new DocumentReference<U>(
+      this.firestore,
+      this._delegate.withConverter(
+        converter as UntypedFirestoreDataConverter<U>
+      )
+    );
   }
+}
 
-  /**
-   * Converts a ViewSnapshot that contains the current document to a
-   * DocumentSnapshot.
-   */
-  private _convertToDocSnapshot(snapshot: ViewSnapshot): DocumentSnapshot<T> {
-    debugAssert(
-      snapshot.docs.size <= 1,
-      'Too many documents returned on a document query'
-    );
-    const doc = snapshot.docs.get(this._key);
+/**
+ * Replaces the function name in an error thrown by the firestore-exp API
+ * with the function names used in the classic API.
+ */
+function rewriteError(
```

Comment:
Since this is specific to method references maybe call this `rewriteErrorMethod` or maybe even just `replaceMethod`? Alternatively, leave it as a general `rewriteError` and make the caller supply the parens?

Reply:
Renamed to `replaceFunctionName`.